src/input_formatter/expression_formatter.ts

```diff
@@ -0,0 +1,155 @@
+			const element: string = replaceASingleExpressionElementIntoPrologCode(preFormattedExpression[index]);
+			finalFormattedExpression.push(element);
+		} else {
+			finalFormattedExpression.push(preFormattedExpression[index]);
+		}
+		index++;
+	}
+	return finalFormattedExpression;
+
+}
+
+/**
+ * helper function for replaceExpressionElementsIntoPrologCode
+ * @return finalExpressionElement
+ */
+export function replaceASingleExpressionElementIntoPrologCode(preformattedExpressionElement: string) {
+	const finalExpression: string = preformattedExpressionElement;
+	return finalExpression
+		.replace(/(bracketLeft)/g, "\[")
+		.replace(/(bracketRight)/g, "\]")
+		.replace(/(equivalence)/g, "<->")
+		.replace(/(implicationRight)/g, "->")
+		.replace(/(implicationLeft)/g, "<-")
+		.replace(/(negation)/g, "neg")
+		.replace(/(conjunction)/g, "and")
+		.replace(/(disjunction)/g, "or")
+		.replace(/(equal)/g, "=")
+		.replace(/(addition)/g, "+")
+		.replace(/(subtraction)/g, "-")
+		.replace(/(multiplication)/g, "*")
+		.replace(/(division)/g, ":")
+		;
+}
+
+export function expressionIssueDetector(preFormattedExpression: string[]) {
+	detectBracketIssues(preFormattedExpression);
+	detectMissingStatementsOrConnector(preFormattedExpression);
+}
+
+/**
+ * Check the expression for an equal amount of opened and closed brackets.; it can detect wrong input
+ */
+export function detectBracketIssues(formattedExpression: string[]) {
+	let bracketCount = 0;
+	let index = 0;
+	while (index < formattedExpression.length && bracketCount >= 0) {
+
+		if (formattedExpression[index].match(/(\[|bracketLeft)/)) {
+			bracketCount++;
+		} else if (formattedExpression[index].match(/(\]|bracketRight)/)) {
+			bracketCount--;
+		}
+		index++;
+	}
+	if (bracketCount > 0) {
+		addIssue("BRACKET_UNDERCLOSING");
+	} else if (bracketCount < 0) {
+		addIssue("BRACKET_OVERCLOSING");
+	}
+}
+
+/**
+ * check the expression for correctness that no logic operators or statements is forgotten; it can detect wrong input
+ */
+export function detectMissingStatementsOrConnector(formattedExpression: string[]) {
+	let index = 0;
+	let foundConnector = true;
+	let foundStatement = false;
+	let foundNegation = false;
+	while (index < formattedExpression.length) {
+		if (formattedExpression[index].match(bracket)) {
+			index++;
+		} else if (formattedExpression[index].match(negation)) {
+			foundNegation = true;
+			index++;
+		} else if (formattedExpression[index].match(logicConnector)) {
+			if (foundNegation === true) {
+				addIssue("MISSING_STATEMENT_AFTER_NEGATION");
+			} else if (foundConnector === true) {
+				addIssue("MISSING_STATEMENT_INSIDE");
+			}
+			foundConnector = true;
+			foundStatement = false;
+			foundNegation = false;
+			index++;
+		} else {
+			if (foundStatement === true) {
+				addIssue("MISSING_CONNECTOR");
+			}
+			foundConnector = false;
+			foundStatement = true;
+			foundNegation = false;
+			index++;
+		}
+	}
```

Comment:
for- or foreach-loop.

Reply:
it had a reason why i used a while loop, because with a index i can count faster through one loop after i matched an if or else if case

Reply:
But you do see that it's pretty easy to build a non-terminating loop here? Are you sure you covered all cases and this code will not loop indefinitly for some input? If you have to use a while-loop somewhere, it's your job to make it clear for readers that it cannot not stop.

Reply:
yes im sure because in each loop each if/ if else/else case has its own index increment line so its impossible not to reach the breakout condition.

Reply:
where is the difference to
```typescript
foreach (const expr of formattedExpression) {
		if (expr.match(bracket)) {
			continue;
		} else if (expr.match(negation)) {
			foundNegation = true;
			continue;
		} else if (expr.match(logicConnector)) {
			if (foundNegation === true) {
				addIssue("MISSING_STATEMENT_AFTER_NEGATION");
			} else if (foundConnector === true) {
				addIssue("MISSING_STATEMENT_INSIDE");
			}
			foundConnector = true;
			foundStatement = false;
			foundNegation = false;
			continue;
		} else {
			if (foundStatement === true) {
				addIssue("MISSING_CONNECTOR");
			}
			foundConnector = false;
			foundStatement = true;
			foundNegation = false;
			continue;
		}
	}
```
The `continue`-statements are not even necessary, but this way the first case isn't left empty.